controller/backing_image_controller.go

```diff
@@ -0,0 +1,605 @@
+		if len(pods) == 0 {
+			log.Info("The data in all disks is removed, then the finalizer will be removed")
+			return bic.ds.RemoveFinalizerForBackingImage(backingImage)
+		}
+		return nil
+	}
+
+	existingBackingImage := backingImage.DeepCopy()
+	defer func() {
+		if err == nil && !reflect.DeepEqual(existingBackingImage.Status, backingImage.Status) {
+			_, err = bic.ds.UpdateBackingImageStatus(backingImage)
+		}
+		if apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debugf("Requeue %v due to conflict", key)
+			bic.enqueueBackingImage(backingImage)
+			err = nil
+		}
+	}()
+
+	if err := bic.syncBackingImageWithDownloadPods(backingImage); err != nil {
+		return err
+	}
+
+	if err := bic.updateDiskLastReferenceMap(backingImage); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getDownloadPodName(biName, diskUUID string) string {
+	return fmt.Sprintf("%s-download-%s", biName, diskUUID[:8])
+}
+
+func (bic *BackingImageController) syncBackingImageWithDownloadPods(backingImage *longhorn.BackingImage) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to sync backing image with download pods")
+	}()
+	log := getLoggerForBackingImage(bic.logger, backingImage)
+
+	if backingImage.Status.DiskDownloadStateMap == nil {
+		backingImage.Status.DiskDownloadStateMap = map[string]types.BackingImageDownloadState{}
+	}
+
+	// Check the records exist only in backingImage.Status.DiskDownloadStateMap
+	for uuid := range backingImage.Status.DiskDownloadStateMap {
+		if _, exists := backingImage.Spec.Disks[uuid]; exists {
+			continue
+		}
+		downloadPodName := getDownloadPodName(backingImage.Name, uuid)
+		downloadPod, err := bic.ds.GetPod(downloadPodName)
+		if err != nil {
+			return err
+		}
+		if downloadPod == nil {
+			log.WithField("diskUUID", uuid).Debugf("Cleaned up the backing image in disk")
+			delete(backingImage.Status.DiskDownloadStateMap, uuid)
+			continue
+		}
+		if downloadPod.DeletionTimestamp == nil {
+			log.WithFields(logrus.Fields{"diskUUID": uuid, "pod": downloadPodName}).Debugf("Deleting pod")
+			if err := bic.ds.DeletePod(downloadPodName); err != nil {
+				return err
+			}
+		}
+		backingImage.Status.DiskDownloadStateMap[uuid] = types.BackingImageDownloadStateTerminating
+	}
+
+	for diskUUID := range backingImage.Spec.Disks {
+		downloadPodName := getDownloadPodName(backingImage.Name, diskUUID)
+		downloadPod, err := bic.ds.GetPod(downloadPodName)
+		if err != nil {
+			return err
+		}
+
+		// Create/Recreate the download pod for a disk if the old pod is gone
+		// and the disk on the node are ready
+		if downloadPod == nil {
+			if state, exists := backingImage.Status.DiskDownloadStateMap[diskUUID]; exists {
+				log.WithField("diskUUID", diskUUID).Warnf("The disk download state is %v but the download pod is nil, will directly create a new pod then",
+					state)
+			}
+			if err := bic.createDownloadPod(downloadPodName, diskUUID, backingImage); err != nil {
```

Comment:
consider returning the created pod and just assigning it to the download pod, then you can remove the continue below, since the pod processing will set it to downloading based on the `pod.pending`.